File: Java/035.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Problem035 {

    private static final Set<Integer> primeSet = new HashSet<>();

    public static boolean[] primeSieve(int n) {
        boolean[] sieve = new boolean[n];
        Arrays.fill(sieve, true);
        for (int i = 2; i * i < n; i++) {
            if (sieve[i]) {
                for (int j = i * i; j < n; j += i) {
                    sieve[j] = false;
                }
            }
        }
        return sieve;
    }

    public static boolean circularPrime(int n) {
        String s = String.valueOf(n);
        for (int i = 0; i < s.length(); i++) {
            if (!primeSet.contains(Integer.parseInt(s.substring(i) + s.substring(0, i)))) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        boolean[] primes = primeSieve(1_000_000);
        for (int i = 2; i < primes.length; i++) {
            if (primes[i]) {
                primeSet.add(i);
            }
        }

        int res = 0;
        for (int prime : primeSet) {
            if (circularPrime(prime)) {
                res++;
            }
        }
        System.out.println(res);
    }
}

File: Python/035.py
# The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.
# There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.
# How many circular primes are there below one million?


import math


def prime_sieve(n):
    """Use the Sieve of Eratosthenes to produce all prime numbers from 2 through n.
    Adapated from the wikipedia pseudocode."""

    sieve = [True] * n

    for i in range(2, int(math.sqrt(n)) + 1):
        if sieve[i]:
            for j in range(2 * i, n, i):
                sieve[j] = False

    return [x for x in range(2, n) if sieve[x]]


primes = prime_sieve(1000000)
prime_set = set(primes)


def circular_prime(n):
    """Checks if the prime n is a circular prime."""
    s = str(n)
    for idx in range(len(s)):
        if int(s[idx:] + s[:idx]) not in prime_set:
            return False
    return True


if __name__ == "__main__":
    res = 0
    for prime in primes:
        if circular_prime(prime):
            res += 1
    print(res)
